commands.go

```diff
@@ -0,0 +1,746 @@
+package ravendb
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+)
+
+// JSONArrayResult represents result of BatchCommand, which is array of JSON objects
+// it's a type alias so that it doesn't need casting when json marshalling
+type JSONArrayResult = []ObjectNode
+
+// RavenCommand represents data needed to issue an HTTP command to the server
+type RavenCommand struct {
+	Method        string // GET, PUT etc.
+	IsReadRequest bool
+	// to create a full url, replace {url} and {db} with ServerNode.URL and
+	// ServerNode.Database
+	URLTemplate string
+	// additional HTTP request headers
+	Headers map[string]string
+	Data    []byte
+
+	failedNodes []*ServerNode
+}
+
+func (c *RavenCommand) addFailedNode(node *ServerNode) {
+	c.failedNodes = append(c.failedNodes, node)
+}
+
+func (c *RavenCommand) isFailedWithNode(node *ServerNode) bool {
+	for _, n := range c.failedNodes {
+		if n == node {
+			return true
+		}
+	}
+	return false
+}
+
+// BuildFullURL creates a full url by filling in server address and database name
+func (c *RavenCommand) BuildFullURL(n *ServerNode) string {
+	url := strings.Replace(c.URLTemplate, "{url}", n.URL, -1)
+	return strings.Replace(url, "{db}", n.Database, -1)
+}
+
+// CommandExecutorFunc takes RavenCommand, sends it over HTTP to the server and
+// returns raw HTTP response
+type CommandExecutorFunc func(cmd *RavenCommand) (*http.Response, error)
+
+// ExecuteCommand executes RavenCommand with a given executor function
+func ExecuteCommand(exec CommandExecutorFunc, cmd *RavenCommand) (*http.Response, error) {
+	return exec(cmd)
+}
+
+func decodeJSONFromReader(r io.Reader, v interface{}) error {
+	return json.NewDecoder(r).Decode(v)
+}
+
+func makeHTTPRequest(n *ServerNode, cmd *RavenCommand) (*http.Request, error) {
+	url := cmd.BuildFullURL(n)
+	var body io.Reader
+	if cmd.Method == http.MethodPut || cmd.Method == http.MethodPost || cmd.Method == http.MethodDelete {
+		// TODO: should this be mandatory?
```

Comment:
No, it isn't mandatory. There are cases where we just `POST` and not have any body